File: Exercism/c/binary-search/src/binary_search.c
#include "binary_search.h"
#include <stddef.h>

int *binary_search(int needle, int haystack[], size_t length) {
  if (haystack != NULL) {
    int left = 0, right = length - 1;
    while (left <= right) {
      int middle = left + ((right - left) / 2);
      if (needle < haystack[middle]) {
        right = middle - 1;
      }
      else if (needle > haystack[middle]) {
        left = middle + 1;
      }
      else {
        return &haystack[middle];
      }
    }
  }
  return NULL;
}

File: Exercism/python/binary-search/binary_search.py
def find(search_list, value):
    # Not great since deep recursion would be a problem, but can be easily
    # changed to use a loop instead.

    def _find_recursive(left, right):
        if left <= right:
            middle = (left + right) // 2
            if value < search_list[middle]:
                return _find_recursive(left, middle - 1)
            elif value > search_list[middle]:
                return _find_recursive(middle + 1, right)
            else:
                return middle
        else:
            raise ValueError("Value not in list")

    return _find_recursive(0, len(search_list) - 1)
